src/platform/error_abi.cpp

```diff
@@ -0,0 +1,191 @@
+#include "pal_internal.h"
+#include "pal_error.h"
+#include "atomic_ref_count.h"
+#include <xlang/base.h>
+
+namespace xlang::impl
+{
+    struct error_info : xlang_error_info
+    {
+        // Used to construct constant errors used in out of memory scenarios.
+        explicit error_info(xlang_result result) noexcept :
+            m_result{ result },
+            m_modifiable{ false }
+        {
+        }
+
+        explicit error_info(
+            xlang_result result,
+            xlang_string message,
+            xlang_string projection_identifier,
+            xlang_string language_error,
+            xlang_unknown* execution_trace,
+            xlang_unknown* language_information
+        ) noexcept :
+            m_result{ result },
+            m_message{ message },
+            m_projection_identifier{ projection_identifier },
+            m_language_error{ language_error }
+        {
+            m_execution_trace.copy_from(execution_trace);
+            m_language_information.copy_from(language_information);
+        }
+
+        int32_t XLANG_CALL QueryInterface(xlang_guid const& id, void** object) noexcept final
+        {
+            if (id == xlang_unknown_guid)
+            {
+                static_assert(std::is_base_of_v<xlang_unknown, xlang_error_info>, "Can only combine these two cases if this is true.");
+                *object = static_cast<xlang_unknown*>(this);
+            }
+            else if (id == xlang_error_info_guid)
+            {
+                *object = static_cast<xlang_error_info*>(this);
+            }
+            else
+            {
+                *object = nullptr;
+                return xlang_hresult_no_interface;
+            }
+            AddRef();
+            return 0;
+        }
+
+        uint32_t XLANG_CALL AddRef() noexcept final
+        {
+            return ++m_count;
+        }
+
+        uint32_t XLANG_CALL Release() noexcept final
+        {
+            auto result = --m_count;
+            if (result == 0)
+            {
+                delete this;
+            }
+            return result;
+        }
+
+        void GetError(xlang_result* error) noexcept override
```

Comment:
feels odd to have the type be snake_case and the methods be PascalCase. I get it for GetUnk/AddRef/Release, but should we standardize on snake_case for everything else?

Reply:
Yea I was thinking about that.  Wasn't sure whether we were going to have a convention for interfaces.  But, if we are fine with special casing IUnknown, I will update these functions.

Reply:
I would use snake_case and add a comment to QI/AddRef/Release indicating why they have different naming convention

Reply:
We probably need to come to an agreement on what to use for interfaces and make sure we are consistent through out as there is other new interfaces where we currently use pascal casing.  Opened #363 to track this.